docs/architecture/adr-065-custom-event-indexing.md

```diff
@@ -0,0 +1,341 @@
+    // only index events with a non-empty type
+    if len(event.Type) == 0 {
+      continue
+    }
+
+    for _, attr := range event.Attributes {
+      if len(attr.Key) == 0 {
+        continue
+      }
+
+      // index iff the event specified index:true and it's not a reserved event
+      compositeKey := fmt.Sprintf("%s.%s", event.Type, string(attr.Key))
+      if compositeKey == types.BlockHeightKey {
+        return fmt.Errorf("event type and attribute key \"%s\" is reserved; please use a different key", compositeKey)
+      }
+
+      if attr.GetIndex() {
+        sqlStmt = sqlStmt.Values(compositeKey, string(attr.Value), h.Header.Height, BlockEventTypeBeginBlock)
+      }
+    }
+  }
+
+  // index end_block events...
+  // execute sqlStmt db query...
+}
+
+func (es *PSQLEventSink) IndexTxEvents(txr *abci.TxResult) error {
+  sqlStmtEvents := sq.Insert("tx_events").Columns("key", "value", "height", "hash", "tx_result_id")
+  sqlStmtTxResult := sq.Insert("tx_results").Columns("tx_result")
+
+
+  // store the tx result
+  txBz, err := proto.Marshal(txr)
+  if err != nil {
+    return err
+  }
+
+  sqlStmtTxResult = sqlStmtTxResult.Values(txBz)
+
+  // execute sqlStmtTxResult db query...
+
+  // index the reserved height and hash indices
+  hash := types.Tx(txr.Tx).Hash()
+  sqlStmtEvents = sqlStmtEvents.Values(types.TxHashKey, hash, txr.Height, hash, txrID)
+  sqlStmtEvents = sqlStmtEvents.Values(types.TxHeightKey, txr.Height, txr.Height, hash, txrID)
+
+  for _, event := range result.Result.Events {
+    // only index events with a non-empty type
+    if len(event.Type) == 0 {
+      continue
+    }
+
+    for _, attr := range event.Attributes {
+      if len(attr.Key) == 0 {
+        continue
+      }
+
+      // index if `index: true` is set
+      compositeTag := fmt.Sprintf("%s.%s", event.Type, string(attr.Key))
+			
+      // ensure event does not conflict with a reserved prefix key
+      if compositeTag == types.TxHashKey || compositeTag == types.TxHeightKey {
+        return fmt.Errorf("event type and attribute key \"%s\" is reserved; please use a different key", compositeTag)
+      }
+		
+      if attr.GetIndex() {
+        sqlStmtEvents = sqlStmtEvents.Values(compositeKey, string(attr.Value), txr.Height, hash, txrID)
+      }
+    }
+  }
+
+  // execute sqlStmtEvents db query...
+}
+
+func (es *PSQLEventSink) SearchBlockEvents(ctx context.Context, q *query.Query) ([]int64, error) {
+  sqlStmt = sq.Select("height").Distinct().From("block_events").Where(q.String())
+
+  // execute sqlStmt db query and scan into integer rows...
+}
+
+func (es *PSQLEventSink) SearchTxEvents(ctx context.Context, q *query.Query) ([]*abci.TxResult, error) {
+  sqlStmt = sq.Select("tx_result_id").Distinct().From("tx_events").Where(q.String())
+
+  // execute sqlStmt db query and scan into integer rows...
+  // query tx_results records and scan into binary slice rows...
+  // decode each row into a TxResult...
+}
+```
+
+### Configuration
```

Comment:
Probably not needed, but want to get thoughts on allowing node operators to inject their on EventSink type?

Reply:
For in-process nodes, I don't see why not.

Reply:
How do you "inject" types? There needs to be a custom built sink implementation for any non-existing type.